src/docfx/lib/msalias/MicrosoftAliasCache.cs

```diff
@@ -0,0 +1,133 @@
+
+namespace Microsoft.Docs.Build
+{
+    internal class MicrosoftAliasCache : IDisposable
+    {
+        private readonly string _cachePath;
+        private readonly double _expirationInHours;
+        private readonly SemaphoreSlim _syncRoot = new SemaphoreSlim(1, 1);
+        private readonly MicrosoftGraphAccessor _microsoftGraphAccessor;
+        private Dictionary<string, MicrosoftAlias> _aliases = new Dictionary<string, MicrosoftAlias>();
+        private IList<string> _tempInvalidAliases = new List<string>();
+        private bool _needUpdate = false;
+
+        public MicrosoftAliasCache(Config config)
+        {
+            _cachePath = AppData.MicrosoftAliasCachePath;
+            _expirationInHours = config.JsonSchema.MicrosoftAliasCacheExpirationInHours;
+
+            _microsoftGraphAccessor = new MicrosoftGraphAccessor(
+                config.JsonSchema.MicrosoftGraphTenantId,
+                config.JsonSchema.MicrosoftGraphClientId,
+                config.JsonSchema.MicrosoftGraphClientSecret);
+
+            if (File.Exists(_cachePath))
+            {
+                var cacheFile = JsonUtility.Deserialize<MicrosoftAliasCacheFile>(ProcessUtility.ReadFile(_cachePath), _cachePath);
+                _aliases = cacheFile.Aliases.Where(msAlias => msAlias.Expiry >= DateTime.UtcNow).ToDictionary(msAlias => msAlias.Alias);
+            }
+        }
+
+        public Task<(Error error, MicrosoftAlias msAlias)> GetAsync(string alias)
+        {
+            if (string.IsNullOrEmpty(alias))
+            {
+                return new Task<(Error error, MicrosoftAlias msAlias)>(null, null);
+            }
+
+            return Synchronized(GetAsyncCore);
+
+            async Task<(Error error, MicrosoftAlias msAlias)> GetAsyncCore()
+            {
+                if (_aliases.TryGetValue(alias, out var msAlias))
+                {
+                    return (null, msAlias);
+                }
+                else
+                {
+                    if (_tempInvalidAliases.Contains(alias))
+                    {
+                        return (null, null);
+                    }
+
+                    var (error, isValid) = await _microsoftGraphAccessor.ValidateAlias(alias);
+
+                    if (error != null)
+                    {
+                        return (error, null);
+                    }
+
+                    if (isValid)
+                    {
+                        var newMsAlias = new MicrosoftAlias()
+                        {
+                            Alias = alias,
+                            Expiry = NextExpiry(),
+                        };
+
+                        _aliases.Add(alias, new MicrosoftAlias() { Alias = alias, Expiry = NextExpiry() });
+                        _needUpdate = true;
+
+                        return (error, newMsAlias);
+                    }
+                    else
+                    {
+                        _tempInvalidAliases.Add(alias);
```

Comment:
`_tempInvalidAliases` is not saved to disk, thus it'll call msgraph API for each build for invalid alias. Consider adding an `IsInvalid` property to `MicrosoftAlias`.

Reply:
We have considered this issue. We are not going to save illegal aliases. Because if we save the illegal alias to the cache, then if the alias is legal again, we have to wait for the expiration time to end. That is, we can't quickly legalize an alias, such as a new Microsoft account.

Reply:
fair, then can you log a message and add a telemetry each time a graph API has been called, Similar to `GitHubUserCache` line 72,81,82

Reply:
OkOk